### PR TITLE
AP_GPS: add AP_GPS_FEMTO driver for Femtomes GPS

### PR DESCRIPTION
**Pull request by Femtomes**
1.Add AP_GPS_FEMTO driver for Femtomes gps device
2.Add femtomes driver in AP_GPS class

### CODE REVIEW

File: libraries/AP_GPS/AP_GPS_FEMTO.cpp [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+
+#include "AP_GPS.h"
+#include "AP_GPS_FEMTO.h"
+#include <AP_Logger/AP_Logger.h>
+
+extern const AP_HAL::HAL& hal;
+
+#define FEMTO_DEBUGGING 0
+
+#define FEMTO_MSG_ID_UAVGPS 		8001
```
Comment (khancyr) [inline libraries/AP_GPS/AP_GPS_FEMTO.cpp:28]:
```suggestion
static const uint16_t FEMTO_MSG_ID_UAVGPS = 8001;
static const uint8_t FEMTO_MSG_ID_BESTPOS = 42;
```
In private in the class

File: libraries/AP_GPS/AP_GPS_FEMTO.cpp [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+}
+
+bool
+AP_GPS_FEMTO::parse(uint8_t temp)
+{
+    switch (femto_msg.femto_decode_state)
+    {
+        default:
+        case femto_msg_parser_t::PREAMBLE1:
+            if (temp == FEMTO_PREAMBLE1)
```
Comment (khancyr) [inline libraries/AP_GPS/AP_GPS_FEMTO.cpp:94]: miss {}

File: libraries/AP_GPS/AP_GPS_FEMTO.cpp [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+                femto_msg.femto_decode_state = femto_msg_parser_t::PREAMBLE2;
+            femto_msg.read = 0;
+            break;
+        case femto_msg_parser_t::PREAMBLE2:
+            if (temp == FEMTO_PREAMBLE2)
```
Comment (khancyr) [inline libraries/AP_GPS/AP_GPS_FEMTO.cpp:99]:
```suggestion
            if (temp == FEMTO_PREAMBLE2) {
                femto_msg.femto_decode_state = femto_msg_parser_t::PREAMBLE3;
            } else {
                femto_msg.femto_decode_state = femto_msg_parser_t::PREAMBLE1;
            }
```

File: libraries/AP_GPS/AP_GPS_FEMTO.cpp [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+            {
+                femto_msg.femto_decode_state = femto_msg_parser_t::PREAMBLE3;
+            }
+            else
+            {
+                femto_msg.femto_decode_state = femto_msg_parser_t::PREAMBLE1;
+            }
+            break;
+        case femto_msg_parser_t::PREAMBLE3:
+            if (temp == FEMTO_PREAMBLE3)
```
Comment (khancyr) [inline libraries/AP_GPS/AP_GPS_FEMTO.cpp:109]: same a previous

File: libraries/AP_GPS/AP_GPS_FEMTO.cpp [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+            break;
+        case femto_msg_parser_t::HEADERDATA:
+            if (femto_msg.read >= sizeof(femto_msg.header.data)) {
+                Debug("parse header overflow length=%u\n", (unsigned)femto_msg.read);
+                femto_msg.femto_decode_state = femto_msg_parser_t::PREAMBLE1;
+                break;
+            }
+            femto_msg.header.data[femto_msg.read] = temp;
+            femto_msg.read++;
+            if (femto_msg.read >= femto_msg.header.femto_header.headerlength)
```
Comment (khancyr) [inline libraries/AP_GPS/AP_GPS_FEMTO.cpp:136]:
```suggestion
            if (femto_msg.read >= femto_msg.header.femto_header.headerlength) {
```

File: libraries/AP_GPS/AP_GPS_FEMTO.cpp [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+            break;
+        case femto_msg_parser_t::DATA:
+            if (femto_msg.read >= sizeof(femto_msg.data)) {
+                Debug("parse data overflow length=%u msglength=%u\n", (unsigned)femto_msg.read,femto_msg.header.femto_header.messagelength);
+                femto_msg.femto_decode_state = femto_msg_parser_t::PREAMBLE1;
+                break;
+            }
+            femto_msg.data.bytes[femto_msg.read - femto_msg.header.femto_header.headerlength] = temp;
+            femto_msg.read++;
+            if (femto_msg.read >= (femto_msg.header.femto_header.messagelength + femto_msg.header.femto_header.headerlength))
```
Comment (khancyr) [inline libraries/AP_GPS/AP_GPS_FEMTO.cpp:149]:
```suggestion
            if (femto_msg.read >= (femto_msg.header.femto_header.messagelength + femto_msg.header.femto_header.headerlength)) {
```

File: libraries/AP_GPS/AP_GPS_FEMTO.cpp [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+bool
+AP_GPS_FEMTO::process_message(void)
+{
+    uint16_t messageid = femto_msg.header.femto_header.messageid;
+
+    Debug("FEMTO process_message messid=%u\n",messageid);
+
+    check_new_itow(femto_msg.header.femto_header.tow, femto_msg.header.femto_header.messagelength + femto_msg.header.femto_header.headerlength);
+    
+    if (messageid == FEMTO_MSG_ID_UAVGPS) // uavgps
```
Comment (khancyr) [inline libraries/AP_GPS/AP_GPS_FEMTO.cpp:198]:
```suggestion
    if (messageid == FEMTO_MSG_ID_UAVGPS) { // uavgps
```

File: libraries/AP_GPS/AP_GPS_FEMTO.cpp [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+        fill_3d_velocity();
+        state.velocity.z = uav_gps.vel_d_m_s;
+        state.have_vertical_velocity = true;
+
+        state.hdop = (uint16_t) (uav_gps.hdop*100);
+        state.vdop = (uint16_t) (uav_gps.vdop*100);
+
+        _new_uavgps = true;
+    }
+    if (messageid == FEMTO_MSG_ID_BESTPOS) // bestpos
```
Comment (khancyr) [inline libraries/AP_GPS/AP_GPS_FEMTO.cpp:255]:
```suggestion
    if (messageid == FEMTO_MSG_ID_BESTPOS) { // bestpos
```

File: libraries/AP_GPS/AP_GPS_FEMTO.cpp [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+
+        state.rtk_age_ms = bestpos.diffage * 1000;
+
+        _last_vel_time = (uint32_t) femto_msg.header.femto_header.tow;
+
+        _new_bestpos = true;
+    }
+
+    // ensure bestpos and uavgps stay insync
+    if (_new_bestpos && _new_uavgps && _last_vel_time == state.time_week_ms)
```
Comment (khancyr) [inline libraries/AP_GPS/AP_GPS_FEMTO.cpp:267]:
```suggestion
    if (_new_bestpos && _new_uavgps && _last_vel_time == state.time_week_ms) {
```

File: libraries/AP_GPS/AP_GPS_FEMTO.h [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+    uint32_t CRC32Value(uint32_t icrc);
+    uint32_t CalculateBlockCRC32(uint32_t length, uint8_t *buffer, uint32_t crc);
+
+    static const uint8_t FEMTO_PREAMBLE1 = 0xaa;
+    static const uint8_t FEMTO_PREAMBLE2 = 0x44;
+    static const uint8_t FEMTO_PREAMBLE3 = 0x12;
+    
+    static const char* const _initialisation_blob[3];
+   
+    uint32_t crc_error_counter = 0;
```
Comment (khancyr) [inline libraries/AP_GPS/AP_GPS_FEMTO.h:50]: you don't need to init to 0 on ArduPilot, that is done by default, so it is better to remove the zero initialisation to save some flash space

File: libraries/AP_GPS/AP_GPS_FEMTO.h [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+    uint32_t last_injected_data_ms = 0;
+
+    uint8_t _init_blob_index = 0;
+    uint32_t _init_blob_time = 0;
+
+    // do we have new position information?
+    bool _new_bestpos:1;
```
Comment (khancyr) [inline libraries/AP_GPS/AP_GPS_FEMTO.h:57]: :1 is useless on bool

File: libraries/AP_GPS/AP_GPS_FEMTO.h [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+    // do we have new uavgps information?
+    bool _new_uavgps:1;
```
Comment (khancyr) [inline libraries/AP_GPS/AP_GPS_FEMTO.h:59]: same

**Review by khancyr — CHANGES_REQUESTED** [review]
hello,

thanks for this.
Do you have a link where we could buy this GPS? and a document on communication protocol?

The implementation seems nice. I have mark some small corrections to do.
You will also need to squash your commits to have only one and change its message to : 
AP_GPS: add AP_GPS_FEMTO driver for Femtomes GPS

File: libraries/AP_GPS/AP_GPS_FEMTO.cpp [hunk 13]
```diff
@@ -0,0 +1,310 @@
...
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+//  Femtomes GPS driver for ArduPilot.
+//  Code by Rui Zheng <ruizheng@femtomes.com>
+
+
+#include "AP_GPS.h"
+#include "AP_GPS_FEMTO.h"
+#include <AP_Logger/AP_Logger.h>
```
Comment (khancyr) [inline libraries/AP_GPS/AP_GPS_FEMTO.cpp:22]: this import is unused

**Comment by Femtomes** [comment]
> hello,
> 
> thanks for this.
> Do you have a link where we could buy this GPS? and a document on communication protocol?
> 
> The implementation seems nice. I have mark some small corrections to do.
> You will also need to squash your commits to have only one and change its message to :
> AP_GPS: add AP_GPS_FEMTO driver for Femtomes GPS

File: libraries/AP_GPS/AP_GPS_FEMTO.cpp (hunk 13 above)
Comment (Femtomes) [reply libraries/AP_GPS/AP_GPS_FEMTO.cpp:22]: ok，I will move it to FEMTO_DEBUGGING place

File: libraries/AP_GPS/AP_GPS_FEMTO.cpp (hunk 1 above)
Comment (Femtomes) [reply libraries/AP_GPS/AP_GPS_FEMTO.cpp:28]: ok

File: libraries/AP_GPS/AP_GPS_FEMTO.cpp (hunk 2 above)
Comment (Femtomes) [reply libraries/AP_GPS/AP_GPS_FEMTO.cpp:94]: ok，I will add {}

**Comment by khancyr** [comment]
You can keep the pr open and force push on it when you finish the modifications !

**Comment by Femtomes** [comment]
Hi, khancyr  
I'm so glad to hear from you,thank you for your trust and support.
Currently we havn't sale on the platform yet, if necessary if you need you can send your request to the e-mail address sales@femtomes.com,where we will reply to you once we receive your letter. We will also provide the link as soon as possible.
I hope we can cooperate more,if possible,you can also be our agent.
Thanks again for your support.
Best regards.

**Comment by Femtomes** [comment]
> hello,
> 
> thanks for this.
> Do you have a link where we could buy this GPS? and a document on communication protocol?
> 
> The implementation seems nice. I have mark some small corrections to do.
> You will also need to squash your commits to have only one and change its message to :
> AP_GPS: add AP_GPS_FEMTO driver for Femtomes GPS

How to squash commits to have only one ?

**Review by WickedShell — CHANGES_REQUESTED** [review]
This appears to be a copy of the Novatel driver (including all the bad habits it has) with 1 additional new message. If it's just one new message I'd strongly prefer to see us parameterize the novatel driver to indicate which we are parsing. I'd like to see more data about the GPS unit, and an actual protocol manual.

**Comment by Femtomes** [comment]
Hi,Wickedshell
Thank you very much for your comments. It means a lot to us.
Currently the drivers utlizes the novatel style BESTPOS due to compatibility issues.
Could you please tell me more about the bad habits you mentioned above , if possible, and any suggestions? 
Meanwhile, we are planing to introduce a new message named UAVSTATUS, which combined with the UAVGPS to provide all the information APM need. The R&D work are also complete. Your comments and suggestions are warmly welcome. 
If you would like to see more data about the GPS unit, and an actual protocol manual, please kindly give an e-mail address , where we can provide a very detailed information and communicate smoother for each other.
Look forward to your feedback and suggestions soon.

**Comment by WickedShell** [comment]
@Femtomes The driver is actually so similiar I'd prefer to see it lumped into the Novatel driver with a parameter passed to the constructor to tell us which type we are actually parsing. That would eliminate a lot of code duplication and reuse.

You can send me an e-mail at `wicked.shell.scripts at gmail.com`

**Comment by tridge** [comment]
@Femtomes the reason for trying to combine your driver with the NOVA driver is to save flash space. We're quite low on flash, and if we can combine the two then we can include the driver with more boards. It would still have a separate GPS_TYPE, so as far as users are concerned it would be a separate driver.
Can you look at the feasibility of combining the functionality with the NOVA driver?

**Pull request closed without merging**